wallet-api/src/Wallet/Emulator/Types.hs

```diff
@@ -0,0 +1,138 @@
+    handleNotifications :: [Notification] -> m ()
+
+data FailureReason = FailureReason T.Text
+    deriving Show
+
+-- Emulator code
+
+-- This is so that assertions can tell whether and why things failed. They’re not visible through the wallet api, but only to assertions, which is helpful for testing
+data FailedTxn = FailedTxn Txn FailureReason
+    deriving Show
+type FailedTxPool = [FailedTxn]
+
+type Assertion = EmulatorState -> Maybe AssertionError
+data AssertionError = AssertionError T.Text
+    deriving Show
+
+isValidated :: Txn -> Assertion
+isValidated txn emState =
+    if notElem txn (join $ emChain emState) then Just $ AssertionError $ "Txn not validated: " <> T.pack (show txn) else Nothing
+
+-- | The type of events in the emulator. @n@ is the type (usually a monad) in which wallet actions
+-- take place.
+data Event n a where
+    -- these need the coordination code to tell us what to do
+    WalletAction :: Agent -> n () -> Event n [Txn]
+    -- these are always responded to in a routine way, check triggers, update wallet state etc.
+    WalletRecvNotification :: Agent -> [Notification] -> Event n [Txn]
+    BlockchainActions :: Event n Block
+    Assertion :: Assertion -> Event n ()
+
+-- Program is like Free, except it makes the Functor for us so we can have a nice GADT
+type Trace = Op.Program (Event EmulatedWalletApi)
+
+data EmulatorState = EmulatorState { emChain :: Chain, emTxPool :: TxPool, emFailedTxPool :: FailedTxPool, emAgentState :: Map Agent WalletState }
+
+emptyEmulatorState :: EmulatorState
+emptyEmulatorState = EmulatorState { emChain = [], emTxPool = [], emFailedTxPool = [], emAgentState = Map.empty }
+
+type MonadEmulator m = (MonadState EmulatorState m, MonadError AssertionError m)
+
+-- TODO: actual validation
+validate :: (MonadEmulator m) => Txn -> m (Either FailedTxn Txn)
+validate txn = pure $ Right txn
+
+liftEmulatedWallet :: (MonadEmulator m) => Agent -> EmulatedWalletApi a -> m ([Txn], a)
+liftEmulatedWallet agent act = do
+    emState <- get
+    let walletState = fromMaybe emptyWalletState $ Map.lookup agent $ emAgentState emState
+    let ((out, newState), txns) = runWriter $ runStateT (runEmulatedWalletApi act) walletState
+    put emState {
+        emTxPool = txns ++ emTxPool emState,
+        emAgentState = Map.insert agent newState $ emAgentState emState
+        }
+    pure (txns, out)
+
+eval :: (MonadEmulator m) => Event EmulatedWalletApi a -> m a
+eval = \case
+    WalletAction agent action -> fst <$> liftEmulatedWallet agent action
+    WalletRecvNotification agent trigger -> fst <$> liftEmulatedWallet agent (handleNotifications trigger)
+    BlockchainActions -> do
+        emState <- get
+        processed <- forM (emTxPool emState) validate
+        let (failed, validated) = partitionEithers processed
+        let block = validated
+        put emState {
+            emChain = block : emChain emState,
+            emFailedTxPool = failed ++ emFailedTxPool emState
```

Comment:
Why do you keep failed transactions? For comparison in tests? (The real blockchain just drops them on the floor.)

Reply:
Yes, exactly (there's a brief note on `FailedTxn`). You might want to assert that a txn failed to validate - you could equate that with saying it's not in the chain or pool, but it seems helpful to have a reason too.

Actually, I think that's premature design, let's drop it for now.